spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -1699,34 +1699,70 @@ struct ProductionVisitor
     }
 
     void operator()(const production::Skip& p) {
+        auto consumeFixedSize = [&](const Expression& size) {
+            assert(size.type().template isA<type::UnsignedInteger>());
+
+            auto n = builder()->addTmp("skip", size);
+            auto loop = builder()->addWhile(builder::greater(n, builder::integer(0U)));
+            pushBuilder(loop, [&]() {
+                pb->waitForInput(builder::integer(1U), "not enough bytes for skipping", p.location());
+                auto consume = builder()->addTmp("consume", builder::min(builder::size(state().cur), size));
+                pb->advanceInput(consume);
+                builder()->addAssign(n, builder::difference(n, consume));
+                builder()->addDebugMsg("spicy-verbose", "- skipped %u bytes (%u left to skip)", {consume, n});
+            });
+        };
+
+        auto fixedSize = [&](const production::Skip& p) {
+            if ( const auto& size = AttributeSet::find(p.field().attributes(), "&size") )
+                return std::optional(size.value().valueAsExpression()->get());
+
+            else if ( auto t = p.field().parseType().tryAs<type::SignedInteger>() )
+                return std::optional(builder::integer(static_cast<uint64_t>(t->width() / 8)));
+
+            else if ( auto t = p.field().parseType().tryAs<type::UnsignedInteger>() )
+                return std::optional(builder::integer(static_cast<uint64_t>(t->width() / 8)));
+
+            else if ( auto t = p.field().parseType().tryAs<type::Bitfield>() )
+                return std::optional(builder::integer(static_cast<uint64_t>(t->width() / 8)));
+
+            else if ( p.field().parseType().isA<type::Address>() ) {
+                if ( AttributeSet::find(p.field().attributes(), "&ipv4") )
+                    return std::optional(builder::integer(4U));
+                else if ( AttributeSet::find(p.field().attributes(), "&ipv6") )
+                    return std::optional(builder::integer(16U));
+                else
+                    hilti::rt::cannot_be_reached();
+            }
+
+            else if ( auto t = p.field().parseType().tryAs<type::Real>() ) {
+                auto type = *p.field().attributes()->find("&type")->valueAsExpression();
+                return std::optional(
+                    builder::ternary(builder::equal(type, builder::id("spicy::RealType::IEEE754_Single")),
+                                     builder::integer(4U), builder::integer(8U)));
+            }
+
```

Comment:
How about making this a type property instead: we could add a API method to types that returns how many bytes their unpacking requires if that's a fixed amount. Same method could then be used by the grammar builder to decide if a `Skip` production can be created. That way we centralize the knowledge with the types themselves, which is easier to maintain.

Reply:
The only types here where we know their sizes are the integers, but the size of everything else depends on some field attribute. I'll try to instead add a method to `Field`.